account-settings: Move disabled change email tooltip to button.

laurynmm opened the pull request:
When an organization disables email changes, we deactivate the change email button for non-admin users in the **Personal Account & Privacy** tab.

Currently, we show an icon when the button is disabled so that users can hover and see a tooltip with information about why their email cannot be changed. This removes the icon and moves the hover tooltip to the disabled button itself.

**Notes**:
- There is a prep commit that simplifies the logic for disabling the button in the handlebars template.
- I know that disabling buttons is not great for accessibility. I'm not sure if moving the tooltip to the deactivated button has a noticeable negative consequence on that front or if it's the same.
- I moved the tooltip from the default "tippy-zulip-tooltip" to a specific case in `static/js/tippy.js` so that the live update behavior for the tooltip would still work if/when an admin changed the setting.

---

**Screenshots and screen captures:**

<details>
<summary>Disabled email changes - hovering</summary>

![Screenshot from 2022-12-09 18-17-46](https://user-images.githubusercontent.com/63245456/206759580-9b7e03fc-88ae-4c37-b0da-f50e230852bc.png)
</details>
<details>
<summary>Disabled email changes - not hovering</summary>

![Screenshot from 2022-12-09 18-17-50](https://user-images.githubusercontent.com/63245456/206759585-a77db81f-12d6-4176-8c1a-79916a0e1d35.png)
</details>
<details>
<summary>When email changes are not disabled - no tooltip on hover</summary>

![Screenshot from 2022-12-09 18-18-20](https://user-images.githubusercontent.com/63245456/206759588-76f71b68-b863-46bb-9d3f-730a7ab3ca12.png)
</details>
<details>
<summary>Non-admin user (left) compared to admin user (right)</summary>

![Screenshot from 2022-12-09 18-20-02](https://user-images.githubusercontent.com/63245456/206759594-c35dc681-8ece-463a-89c0-0e79a80f6df7.png)
</details>

---

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [ ] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/version-control.html)).

- [ ] Each commit is a coherent idea.
- [ ] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

alya commented:
> I know that disabling buttons is not great for accessibility. I'm not sure if moving the tooltip to the deactivated button has a noticeable negative consequence on that front or if it's the same.

Hm, I hadn't been aware of that consideration. Anyway, skimming through https://www.smashingmagazine.com/2021/08/frustrating-design-patterns-disabled-buttons/, at least this particular source actually recommends having a tooltip on disabled buttons.

alya commented:
Let's trigger the tooltip when hovering anywhere over the email, not just the pencil.

alya commented:
How much work would it be to make the same change for the name field? If it's not super-easy, we can just file an issue for it, referring to this PR as a pattern that can be followed.

![Screen Shot 2022-12-09 at 11 47 18 AM](https://user-images.githubusercontent.com/2090066/206785867-a177177d-842b-4816-8c1c-30565ab790cc.png)

laurynmm commented:
> Let's trigger the tooltip when hovering anywhere over the email, not just the pencil.

It did trigger anywhere on the button when I was testing.

Was it different for you when testing? Or did the screenshots make it appear limited to the pencil?

I'm still having trouble doing screen recording on my work computer, but can try to get that cleared up so that I can post better info on my PRs.

laurynmm commented:
> How much work would it be to make the same change for the name field? If it's not super-easy, we can just file an issue for it, referring to this PR as a pattern that can be followed.

This should be easy to do as a follow-up if the code changes for moving the tooltip here are good. The other similar button I have tagged/noted is the 'Deactivate organization' button. As well as the work on disabling the 'Deactivate account' button (#23797).

<details>
<summary>Deactivate organization</summary>

![Screenshot from 2022-12-12 15-07-46](https://user-images.githubusercontent.com/63245456/207065854-b35708ba-a4ee-49a2-89f7-da5b64f934a2.png)

</details>

---

**Updates**:
- Marked as ready for a code review.
- Reworked the prep commit so that the logic for `realm_email_changes_disabled` is in the same place as similar realm settings, such as `realm_name_changes_disabled` and `realm_avatar_changes_disabled`.

alya commented:
@sahil839 In case you can help debug the hover tooltip behavior.

laurynmm commented:
> That was reported from testing, and I'm still seeing the same issue when I re-test (in Chrome on a Mac). The mouse pointer looks disabled as expected, but the tooltip doesn't show up unless I'm hovering right over the pencil.

Thanks for confirming and specifying the browser; I was able to reproduce in Chrome on Linux.

I've pushed an update that triggers the tooltip on the email address and pencil icon in Chrome by wrapping the email text in a `<span>` element. The tooltip doesn't trigger on the padding of the button where there isn't text/icon in Chrome, but in Firefox it does.

@sahil839 - If you have a chance to do a quick review of these changes, that would be great as I haven't done much with the non-default tippy tooltips.

sahil839 commented:
While testing the previous version (i.e. without adding the `span` element), I noticed if we add `pointer-events: none` for the button element when it is disabled and move `cursor: not-allowed` to the wrapper `div` then, it works as expected. I am not sure why this is the correct solution, but my guess is that it may be something related to event propagation.

sahil839 commented on static/templates/settings/account_settings.hbs:
```diff
@@ -8,12 +8,12 @@
-                        <i class="tippy-zulip-tooltip fa fa-question-circle change_email_tooltip settings-info-icon" {{#if user_can_change_email}}style="display: none;"{{/if}} data-tippy-content="{{t 'Email address changes are disabled in this organization.' }}">
-                        </i>
+                        <div id="change_email_container" {{#unless user_can_change_email}}class="inline-block change_email_tooltip"{{else}}class="inline-block"{{/unless}}>
+                            <button id='change_email' type="button" class="button btn-link small rounded inline-block" {{#unless user_can_change_email}}disabled="disabled"{{/unless}}>
```
Would be better to use double quotes for ID here just for consistency.

laurynmm replied:
When updating this in a prep commit, I also took the opportunity to make the ids more clear / specific.

So, the button id is now `change_email_button` (instead of `change_email`) and the form id for the modal is `change_email_form` (instead of `change_email_container`). 

And the id for the new `<div>` wrapper for the tooltip in the 3rd commit is `change_email_button_container`.

timabbott replied:
Definitely an improvement, thanks!

laurynmm commented:
@sahil839 - Thank you for the review!

**Updates**:
- Added a prep commit that refactors the ids for the button and form for email changes to be more clear / specific.
- Updated the final (3rd) commit (which moves the tooltip to the deactivated button) for the review feedback above.
  - When the change email button is disabled, both the button and the button container for the `email_changes_disabled_tooltip` have `cursor: not-allowed`. The button inherits that from `app_components.css`, while the container's CSS rule is specific to its id and class.
  - Also, removes the CSS rule for the now non-existant `email-change-form` class.
  - Manually tested in both Chrome and Firefox (on Linux) for new hover / live update behavior.

sahil839 commented on static/templates/settings/account_settings.hbs:
```diff
@@ -8,12 +8,12 @@
-                        </button>
-                        <i class="tippy-zulip-tooltip fa fa-question-circle change_email_tooltip settings-info-icon" {{#if user_can_change_email}}style="display: none;"{{/if}} data-tippy-content="{{t 'Email address changes are disabled in this organization.' }}">
-                        </i>
+                        <div id="change_email_button_container" {{#unless user_can_change_email}}class="inline-block email_changes_disabled_tooltip"{{else}}class="inline-block"{{/unless}}>
```
This can instead be `class="inline-block {{#unless user_can_change_email}}email_changes_disabled_tooltip{{/unless}}"`

sahil839 reviewed (commented):
Posted one comment. Otherwise LGTM.

alya commented:
Works for me in manual testing!

@timabbott this PR has been reviewed by @sahil839 and me -- please take a look!

timabbott commented:
Merged, thanks @laurynmm and @sahil839!
